File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
'''
백준 실버 2
트리의 부모 찾기
'''

import sys
input = sys.stdin.readline
from collections import deque
queue = deque()
queue.append(1)
def bfs():
    while queue:
        a = queue.popleft()
        for k in nodes[a]:
            if parents[k] == 0:
                parents[k] = a
                queue.append(k)



n = int(input())
nodes = [[] for _ in range(n+1)]
parents = [0]*(n+1)

for _ in range(n-1):
    a, b = map(int,input().split())
    nodes[a].append(b)
    nodes[b].append(a)

bfs()
for child in parents[2:]:
    print(child)

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const bfs = () => {
  q = [];
  q.push(1);

  while (q.length) {
    const current = q.shift();
    for (child of graph[current]) {
      if (parent[child] == 0) {
        parent[child] = current;
        q.push(child);
      }
    }
  }
};

const node = Number(input[0]);
let graph = Array.from({ length: node + 1 }, () => []);
let parent = Array.from({ length: node + 1 }, () => [0]);

for (let i = 1; i < input.length; i++) {
  const [a, b] = input[i].split(" ").map(Number);
  graph[a].push(b);
  graph[b].push(a);
}

bfs();
for (let i = 2; i < parent.length; i++) {
  console.log(parent[i]);
}
